packages/legacy-client/src/ontology-runtime/ontologyProvider/calls/executeQuery.ts

```diff
@@ -0,0 +1,415 @@
+  LocalDate,
+  Timestamp,
+} from "../../baseTypes";
+import type {
+  ObjectSetDefinition,
+  ParameterValue,
+  PrimitiveParameterValue,
+  QueryBucketKey,
+} from "../../baseTypes";
+import { ExecuteQueryErrorHandler, handleExecuteQueryError } from "..";
+import { isOk } from "../Result";
+import { getObject } from "./getObject";
+import { wrapResult } from "./util/wrapResult";
+
+export function executeQuery<
+  O extends OntologyDefinition<any>,
+  Q extends QueryNamesFrom<O>,
+>(
+  client: ThinClient<O>,
+  apiName: Q,
+  params?: QueryParameters<O, Q>,
+): WrappedQueryReturnType<O, Q> {
+  return wrapResult(
+    async () => {
+      const response: { value: PrimitiveParameterValue } = await executeQueryV2(
+        createOpenApiRequest(client.stack, client.fetch),
+        client.ontology.metadata.ontologyApiName,
+        apiName as string,
+        {
+          parameters: params ? getRemappedParameters(params) : {},
+        },
+      );
+      const remappedResponse = await remapQueryResponseType(
+        client,
+        client.ontology.queries[apiName].output,
+        response.value,
+      );
+
+      return {
+        value: remappedResponse,
+      } as any; // TODO this still needs to get removed
+    },
+    e =>
+      handleExecuteQueryError(
+        new ExecuteQueryErrorHandler(),
+        e,
+        e.parameters,
+      ),
+  );
+}
+
+function getRemappedParameters(
+  params: { [parameterId: string]: any },
+): { [parameterId: string]: any } {
+  const parameterMap: { [parameterName: string]: any } = {};
+  const remappedParams = Object.entries(params).reduce(
+    (acc, [key, value]) => {
+      acc[key] = getParameterValueMapping(value);
+      return acc;
+    },
+    parameterMap,
+  );
+
+  return remappedParams;
+}
+
+function getParameterValueMapping(
+  value: ParameterValue,
+): PrimitiveParameterValue {
+  if (isOntologyObject(value)) {
+    return getParameterValueMapping(value.__primaryKey);
+  } else if (value instanceof LocalDate) {
+    return value.toISOString();
+  } else if (value instanceof Timestamp) {
+    return value.toISOString();
+  } else if (isAttachment(value)) {
+    return value.attachmentRid!;
+  } else if (Array.isArray(value)) {
+    return value.map(a => getParameterValueMapping(a));
+  } else if (value instanceof Set) {
+    return Array.from(value, getParameterValueMapping);
+  } else if (GeoShape.isGeoShape(value)) {
+    return value.toGeoJson();
+  } else if (value instanceof GeoPoint) {
+    return value.toGeoJson();
+  } else if (isOntologyObjectSet(value)) {
+    return value.objectSetDefinition;
+  } else if (typeof value === "object") {
+    // Since structs are valid arguments for Queries, we map the values
+    return Object.entries(value).reduce((acc, [key, structValue]) => {
+      acc[key] = getParameterValueMapping(structValue);
+      return acc;
+    }, {} as { [key: string]: PrimitiveParameterValue });
+  }
+
+  return value as string | number | boolean;
```

Comment:
This return type is a bit of a lie, you can get Array<any>, Record<string, any>, and an ObjectSetDefinition